File: evals/parity/reference/hte.R
# HTE reference recipe (R) — grf causal forest + best linear projection.
# `df` is preloaded by run_parity.py from the shared fixture.
# Fixture is a randomized-treatment design (known propensity 0.5) with linear
# heterogeneity in centered age; compared estimands concentrate across RNGs.
suppressMessages(library(grf))

Y <- df$outcome
W <- df$treatment
X <- as.matrix(df[, c("age", "income", "gender")])

# Center age so the BLP slope targets tau change per year about the sample mean,
# matching the Python LinearDML projection on the same centered features.
Xc <- X
Xc[, "age"] <- Xc[, "age"] - 40

cf <- causal_forest(
  X, Y, W,
  W.hat = rep(0.5, length(Y)),  # known randomized propensity
  num.trees = 4000,
  honesty = TRUE,
  seed = 42
)

# Overall ATE via doubly-robust AIPW scores.
ate <- average_treatment_effect(cf)
cat(sprintf("ATE:%f\n", ate[1]))

# Best linear projection of the CATE onto the (centered) effect modifiers.
blp <- best_linear_projection(cf, Xc)
cat(sprintf("BLP_AGE:%f\n", blp["age", "Estimate"]))

# Heterogeneity sign check: top-minus-bottom CATE-quintile GATE (printed, not gated).
tau <- predict(cf)$predictions
q <- cut(tau, quantile(tau, 0:5 / 5), include.lowest = TRUE, labels = 1:5)
hi <- average_treatment_effect(cf, subset = q == 5)[1]
lo <- average_treatment_effect(cf, subset = q == 1)[1]
cat(sprintf("GATE_HIGH_MINUS_LOW:%f\n", hi - lo))

File: evals/parity/reference/hte.py
# HTE reference recipe (Python) — econml CausalForestDML + LinearDML projection.
# `df` is preloaded by run_parity.py from the shared fixture (matches hte.R).
# Randomized-treatment design (known propensity 0.5) with linear heterogeneity in
# centered age; compared estimands (ATE, BLP_AGE) concentrate across RNGs.
import numpy as np
import statsmodels.api as sm
from econml.dml import CausalForestDML, LinearDML
from sklearn.dummy import DummyClassifier
from sklearn.ensemble import GradientBoostingRegressor

Y = df["outcome"].values
T = df["treatment"].values
X = df[["age", "income", "gender"]].values.astype(float)

# Center age so the LinearDML slope targets tau change per year about the sample
# mean, matching the R grf best_linear_projection on the same centered features.
Xc = X.copy()
Xc[:, 0] = Xc[:, 0] - 40.0

# W (confounders / outcome predictors). Treatment is randomized, so propensity is
# the known constant supplied via a prior DummyClassifier (no propensity model).
W = df[["income", "gender"]].values.astype(float)

# Overall ATE via the causal forest (DML/AIPW).
cf = CausalForestDML(
    model_y=GradientBoostingRegressor(n_estimators=200, max_depth=3, random_state=42),
    model_t=DummyClassifier(strategy="prior"),  # known randomized propensity
    discrete_treatment=True,
    n_estimators=4000,
    cv=5,
    random_state=42,
)
cf.fit(Y, T, X=X, W=W)
ate = float(cf.ate(X))
print(f"ATE:{ate:f}")

# Best linear projection of the CATE onto the (centered) effect modifiers, via the
# orthogonal-moment LinearDML coefficients (apples-to-apples with grf's BLP slope).
ld = LinearDML(
    model_y=GradientBoostingRegressor(n_estimators=200, max_depth=3, random_state=42),
    model_t=DummyClassifier(strategy="prior"),
    discrete_treatment=True,
    cv=5,
    random_state=42,
)
ld.fit(Y, T, X=Xc, W=W)
print(f"BLP_AGE:{ld.coef_[0]:f}")

# Heterogeneity sign check: top-minus-bottom CATE-quintile GATE (printed, not gated).
import pandas as pd
cate = cf.effect(X)
q = pd.qcut(cate, 5, labels=[1, 2, 3, 4, 5])
hi = float(cf.ate(X[q == 5]))
lo = float(cf.ate(X[q == 1]))
print(f"GATE_HIGH_MINUS_LOW:{hi - lo:f}")
